packaging/tags.py

```diff
@@ -0,0 +1,382 @@
+        # methods.
+        path_type = type(path)
+        try:
+            path = path_type.__fspath__(path)
+        except AttributeError:
+            if hasattr(path_type, '__fspath__'):
+                raise
+        else:
+            if isinstance(path, (str, bytes)):
+                return path
+            else:
+                raise TypeError("expected __fspath__() to return str or bytes,"
+                                " not " + type(path).__name__)
+
+        raise TypeError("expected str, bytes or os.PathLike object, not "
+                        + path_type.__name__)
+import os.path
+import platform
+import sys
+import sysconfig
+
+
+INTERPRETER_SHORT_NAMES = {
+    "python": "py",  # Generic.
+    "cpython": "cp",
+    "pypy": "pp",
+    "ironpython": "ip",
+    "jython": "jy",
+}
+
+
+_32_BIT_INTERPRETER = sys.maxsize <= 2 ** 32
+
+
+# A dataclass would be better, but Python 2.7. :(
+class Tag:
+
+    def __init__(self, interpreter, abi, platform):
+        self._tags = interpreter.lower(), abi.lower(), platform.lower()
+
+    def __eq__(self, other):
+        return self._tags == other._tags
+
+    def __hash__(self):
+        return hash(self._tags)
+
+    def __str__(self):
+        return "-".join(self._tags)
+
+    def __repr__(self):
+        return "<{self} @ {self_id}>".format(self=self, self_id=id(self))
+
+    @property
+    def interpreter(self):
+        return self._tags[0]
+
+    @property
+    def abi(self):
+        return self._tags[1]
+
+    @property
+    def platform(self):
+        return self._tags[2]
+
+
+def parse_tag(tag):
+    tags = set()
+    interpreters, abis, platforms = tag.split("-")
+    for interpreter in interpreters.split("."):
+        for abi in abis.split("."):
+            for platform_ in platforms.split("."):
+                tags.add(Tag(interpreter, abi, platform_))
+    return frozenset(tags)
+
+
+def parse_wheel_filename(path):
+    name = os.path.splitext(fspath(path))[0]
+    index = len(name)
+    for _ in range(3):  # interpreter, ABI, platform.
+        index = name.rindex("-", 0, index)
+    return parse_tag(name[index + 1:])
+
+
+def _normalize_string(string):
+    return string.replace(".", "_").replace("-", "_")
+
+
+def _cpython_interpreter(py_version):
+    # TODO: Is using py_version_nodot for interpreter version critical?
+    return "cp{major}{minor}".format(major=py_version[0], minor=py_version[1])
+
+
+def _cpython_abi(py_version):
+    soabi = sysconfig.get_config_var("SOABI")
+    if soabi:
+        _, options, _ = soabi.split("-", 2)
+    else:
+        found_options = [str(py_version[0]), str(py_version[1])]
+        if sysconfig.get_config_var("Py_DEBUG"):
+            found_options.append("d")
+        if sysconfig.get_config_var("WITH_PYMALLOC"):
+            found_options.append("m")
+        if sysconfig.get_config_var("Py_UNICODE_SIZE") == 4:
+            found_options.append("u")
+        options = "".join(found_options)
```

Comment:
pip's version of the SOABI fallback code is more elaborate – if `get_config_var` returns None then it tries to sniff some other way. Is removing those fallbacks intentional?

Reply:
Yes, somewhat. I didn't find a problem with the config vars so I assumed it was a hold-over from pre-2.7 days. So if there's a need to add them back in I can, but I'm trying to modernize this code to have less baggage to make it easier to maintain.

Reply:
I've added a TODO to know whether the simplification is acceptable.